tests/playbooks/tests_states.yml

```diff
@@ -47,3 +47,32 @@
             state: down
     - include_tasks: tasks/assert_device_absent.yml
     - include_tasks: tasks/assert_profile_absent.yml
+
+    #test case (downing a non existent profile)
+    #I can set down and remove a non existent profile
+    #This will be rescued until issue 188 is fixed
+    - name: Verify that the play fails with the expected error message
+      block:
+        - name: Set down non existent profile
+          include_role:
+            name: linux-system-roles.network
+          vars:
+            network_connections:
+              - name: non_existent
+                type: ethernet
+                state: down
+                persistent_state: absent
+
+        - name: 'UNREACH'
+          set_fact:
+            __network_test_failed: false
+
+      rescue:
+        - debug:
+            msg: 'The role has failed'
```

Comment:
I don't think you need the `debug` here anymore.  Do you need the `rescue` block at all, or is the `rescue` block that resets the failed task state for Ansible to continue with subsequent tasks?  Try it and see if it works without the `rescue` block.  If you do need the `rescue` block, then perhaps something like this:
```
       rescue:
         set_fact:
           __network_test_failed: true
```

Reply:
> I don't think you need the `debug` here anymore. Do you need the `rescue` block at all, or is the `rescue` block that resets the failed task state for Ansible to continue with subsequent tasks? 

The `rescue` block is needed so that the failed task is reset and the playbook continues as usual since we need the playbook to succeed even when this task fails (which it will until #188 is fixed) . 

> Try it and see if it works without the `rescue` block. If you do need the `rescue` block, then perhaps something like this:
>
> ```
>        rescue:
>          set_fact:
>            __network_test_failed: true
> ```
Wouldn't setting the rescue block as this mean that the `always` block will succeed and always show  ` success_msg: "The role failed when it should have" ` and we wouldn't know whether the role failed on the correct task ?

Reply:
> > I don't think you need the `debug` here anymore. Do you need the `rescue` block at all, or is the `rescue` block that resets the failed task state for Ansible to continue with subsequent tasks?
> 
> The `rescue` block is needed so that the failed task is reset and the playbook continues as usual since we need the playbook to succeed even when this task fails (which it will until #188 is fixed) .

Ok.

> 
> > Try it and see if it works without the `rescue` block. If you do need the `rescue` block, then perhaps something like this:
> > ```
> >        rescue:
> >          set_fact:
> >            __network_test_failed: true
> > ```
> 
> Wouldn't setting the rescue block as this mean that the `always` block will succeed and always show `success_msg: "The role failed when it should have"` and we wouldn't know whether the role failed on the correct task ?

I'm not sure what you mean.  And perhaps my terminology/variable naming is confusing.  Maybe something like this:
```yaml
    - name: Verify that the play fails with the expected error message
      block:
        - name: Set down non existent profile
          include_role:
            name: linux-system-roles.network
          vars:
            network_connections:
              - name: non_existent
                type: ethernet
                state: down
                persistent_state: absent

        - name: This task should never execute
          set_fact:
            __network_role_failed: false

      rescue:
        - name: Role correctly reported failure
          set_fact:
            __network_role_failed: true

      always:
        - assert:
            that:  __network_role_failed
```
That is, what this `block/rescue/always` block is testing is that the role fails as expected.   And, you can test this, by making the `include_role` task succeed.

> we wouldn't know whether the role failed on the correct task ?

It depends on what you mean.  If by "correct task" you mean some specific task inside the network role, then we have no way to know that since `ansible_failed_task` is not working.  If by "correct task" you mean the `include_role` task in the `block` then yes, we know that this task failed because it is the only task in the `block`, so it is the only task which will go into the `rescue` block upon failure.

If this is not acceptable - if we really need to know exactly which network role task failed, then you'll have to implement something using `ansible_failed_test` that checks the error message and other context available.

Reply:
>It depends on what you mean. If by "correct task" you mean some specific task inside the network role, then we have no way to know that since ansible_failed_task is not working. If by "correct task" you mean the include_role task in the block then yes, we know that this task failed because it is the only task in the block, so it is the only task which will go into the rescue block upon failure.

My bad , I was still thinking about the previous configuration with the `UNREACH` task which was to fail if the `include_role` task did not . I'll change this accordingly .